polkadot/node/core/candidate-validation/src/lib.rs

```diff
@@ -153,106 +154,112 @@ async fn run<Context>(
 	ctx.spawn_blocking("pvf-validation-host", task.boxed())?;
 
+	let mut tasks = FuturesUnordered::new();
+	// The task queue size is chosen to be somewhat bigger than the PVF host incoming queue size
+	// to allow exhaustive validation messages to fall through in case the tasks are clogged with
+	// `ValidateFromChainState` messages awaiting data from the runtime
+	let semaphore = Arc::new(Semaphore::new(polkadot_node_core_pvf::HOST_MESSAGE_QUEUE_SIZE * 2));
+
 	loop {
-		match ctx.recv().await? {
-			FromOrchestra::Signal(OverseerSignal::ActiveLeaves(_)) => {},
-			FromOrchestra::Signal(OverseerSignal::BlockFinalized(..)) => {},
-			FromOrchestra::Signal(OverseerSignal::Conclude) => return Ok(()),
-			FromOrchestra::Communication { msg } => match msg {
-				CandidateValidationMessage::ValidateFromChainState {
-					candidate_receipt,
-					pov,
-					executor_params,
-					exec_timeout_kind,
-					response_sender,
-					..
-				} => {
-					let bg = {
-						let mut sender = ctx.sender().clone();
-						let metrics = metrics.clone();
-						let validation_host = validation_host.clone();
-
-						async move {
-							let _timer = metrics.time_validate_from_chain_state();
-							let res = validate_from_chain_state(
-								&mut sender,
-								validation_host,
-								candidate_receipt,
-								pov,
-								executor_params,
-								exec_timeout_kind,
-								&metrics,
-							)
-							.await;
-
-							metrics.on_validation_event(&res);
-							let _ = response_sender.send(res);
-						}
-					};
-
-					ctx.spawn("validate-from-chain-state", bg.boxed())?;
-				},
-				CandidateValidationMessage::ValidateFromExhaustive {
-					validation_data,
-					validation_code,
-					candidate_receipt,
-					pov,
-					executor_params,
-					exec_timeout_kind,
-					response_sender,
-					..
-				} => {
-					let bg = {
-						let metrics = metrics.clone();
-						let validation_host = validation_host.clone();
-
-						async move {
-							let _timer = metrics.time_validate_from_exhaustive();
-							let res = validate_candidate_exhaustive(
-								validation_host,
-								validation_data,
-								validation_code,
-								candidate_receipt,
-								pov,
-								executor_params,
-								exec_timeout_kind,
-								&metrics,
-							)
-							.await;
-
-							metrics.on_validation_event(&res);
-							let _ = response_sender.send(res);
-						}
-					};
-
-					ctx.spawn("validate-from-exhaustive", bg.boxed())?;
-				},
-				CandidateValidationMessage::PreCheck {
-					relay_parent,
-					validation_code_hash,
-					response_sender,
-					..
-				} => {
-					let bg = {
-						let mut sender = ctx.sender().clone();
-						let validation_host = validation_host.clone();
-
-						async move {
-							let precheck_result = precheck_pvf(
-								&mut sender,
-								validation_host,
-								relay_parent,
-								validation_code_hash,
-							)
-							.await;
-
-							let _ = response_sender.send(precheck_result);
+		futures::select! {
```

Comment:
Why did we gave up on spawn I imagined we can do something like this, what am I missing ?
```
--- a/polkadot/node/core/candidate-validation/src/lib.rs
+++ b/polkadot/node/core/candidate-validation/src/lib.rs
@@ -152,6 +152,8 @@ async fn run<Context>(
        )
        .await;
        ctx.spawn_blocking("pvf-validation-host", task.boxed())?;
+       use async_semaphore::Semaphore;
+       let semaphore = Arc::new(Semaphore::new(10));
 
        loop {
                match ctx.recv().await? {
@@ -171,9 +173,12 @@ async fn run<Context>(
                                                let mut sender = ctx.sender().clone();
                                                let metrics = metrics.clone();
                                                let validation_host = validation_host.clone();
+                                               let guard = semaphore.acquire_arc().await;
 
                                                async move {
                                                        let _timer = metrics.time_validate_from_chain_state();
+                                                       let _guard = guard;
+
                                                        let res = validate_from_chain_state(
                                                                &mut sender,
                                                                validation_host,
@@ -205,9 +210,12 @@ async fn run<Context>(
                                        let bg = {
                                                let metrics = metrics.clone();
                                                let validation_host = validation_host.clone();
+                                               let guard = semaphore.acquire_arc().await;
 
                                                async move {
                                                        let _timer = metrics.time_validate_from_exhaustive();
+                                                       let _guard = guard;
+
                                                        let res = validate_candidate_exhaustive(
                                                                validation_host,
                                                                validation_data,
@@ -236,8 +244,11 @@ async fn run<Context>(
                                        let bg = {
                                                let mut sender = ctx.sender().clone();
                                                let validation_host = validation_host.clone();
+                                               let guard = semaphore.acquire_arc().await;
 
                                                async move {
+                                                       let _guard = guard;
+
```

Reply:
Never mind :D I was dumb we need the select if don't want to block signals processing while the queue is full.

Reply:
> I was dumb we need the select if don't want to block signals processing while the queue is full.

AFAICT, in its current form, this PR will still not process signals while the queue is full. The `select` only helps with the other tasks being processed concurrently with the overseer messages (if we were using spawn, we wouldn't need the select)

Reply:
True, and as far as I understand, that's exactly what we want. When the queue is full, signal processing stops, thus channeling the backpressure upstream.

Reply:
If we'd like to not block signal processing when the queue is full (which is really only for Conclude) we should add a separate task that handles `CandidateValidationMessage`s (and the main task sends work to it via a channel).

Not having this would prevent the subsystem from shutting down if the queue is full I guess. Not sure if it's a real problem with a queue size of 10.. idk

Reply:
side note: maybe we should also have some tests that simulate a full queue, if that's feasible

Reply:
> we should add a separate task that handles `CandidateValidationMessage`s (and the main task sends work to it via a channel).

Hmm, that's interesting, really interesting. CC @eskimor @rphmeier 

Let's consider a separate task that's processing only candidate validation messages. The main subsystem loop sends the messages to that task via a channel. There are two possibilities:
1) If the queue is full, that task ceases to receive new messages. In that case, the main loop `await`s on `send` to that task, and nothing changes in practice; the main loop is still blocked and doesn't process incoming signals.
2) The task swallows all the messages but only processes N validation jobs in parallel. Sounds good, but no backpressure then! That is, that design works exactly as it works right now, just with a limited number of tasks (which is probably better than now), but it doesn't channel any backpressure upstream.

Trying to solve that, I went to see how it works in other subsystems. [Backing main loop `await`s on `handle_communication`](https://github.com/paritytech/polkadot-sdk/blob/6b7be115fd523057d01cea46e9680a62a0b08d97/polkadot/node/core/backing/src/lib.rs#L356), so it doesn't process overseer signals during that handling either. But what happens next is even more interesting. I would expect that at some point, it sends a candidate validation message and `await`s the answer, thus creating backpressure on the backing subsystem. But in reality... Kaboom, [it's spawning a non-blocking task to handle validation](https://github.com/paritytech/polkadot-sdk/blob/6b7be115fd523057d01cea46e9680a62a0b08d97/polkadot/node/core/backing/src/lib.rs#L1713)! And the number of those tasks is unbounded as well, so the backing subsys doesn't even try to create that backpressure I'm trying to channel upstream in this PR!

I haven't had a look at other subsystems yet, but unless I'm missing something, it's not the candidate validation subsystem that makes bounded channels useless. The backpressure thing seems broken at many points.

Reply:
Not processing overseer signal sounds worse than it is. Signal processing is prioritized and there is only a very limited amount. Also the signal channel has size 50 (if I remember correctly). If this ever gets full, you have a bug.

If the task you are blocking on is not blocking for something crazy like 30 seconds, but will always finish at least in the single digit second range, then there should be no problem ever.

Let's assume we get really a lot of signals (10 per slot). This means to get the signal channel full (to actually back pressure), even in this scenario we have 5*6s = 30s time. If your task is faster than this, we will be processing all that queued up signals before processing anything else. Therefore the next time we wait, the signal queue is empty and we have another 30s (pessimistically).

Therefore in practice, assuming signals themselves are not inducing work, we only really back pressure on messages, not signals.

That being said, we could be smarter in back pressure with candidate validation. E.g. the backing subsystem, when it notices a full channel to candidate validation - it could acknowledge that candidate validation is loaded and just drop the message.

Reply:
> If the task you are blocking on is not blocking for something crazy like 30 seconds

It does :sweat_smile: 
Pre-checking requests go the same way, and they have 60-sec hard timeout by default.
So it's still better to make them async. Not merging this one yet, will make some improvements.

Reply:
I think we just need to rewrite in way that we only allow a bounded amount of tasks to be running at a certain time. Also, the semaphore is not the way I would do it. I propose we implement backpressure differently, by returing something like Err(BusyPleaseRetry). This allows us to still process signals,but we'll have to handle this backpressure in backing/approval voting/disputes.

Reply:
Alternatively, we could downgrade `candidate-validation` to be a library rather than subsystem, and then yes, we could use this semaphore approach to limit how many run in parallel and block at caller.

Reply:
This is essentially the same as I am proposing, but with more complexity. Letting the channels go full, does exactly that: "Err(BussyPleaseRetry)" ... you can either retry/drop or block, if you really need it.

What would be the problem adding a call that only receives signals? If I remember correctly the normal recv is a select over signals and messages - should be super trivial to add this.

Reply:
Yeah, that is easy to add

Reply:
@s0me0ne-unkn0wn you wanna take a stab? Apart from finding the code in that macro soup, the actual addition should be trivial.

Reply:
@eskimor no problem, I just need to understand your idea better. Do you mean you want to introduce methods like `ctx.recv_signal()` and `ctx.recv_non_signal()` to be able to process them in parallel? Because if we have normal `ctx.recv()` along with `ctx.recv_signal()`, it looks somewhat redundant, we'll get the same signals in two places.

Also, are there any points where sequential processing of signal and non-signal messages can make a difference? For example, if some message relies on the leaf activation signal being already processed by the subsystem?

Reply:
Maybe a dumb question but how exactly do signals and messages differ? Signals only notify of some change in state (i.e. shutdown, leaf update, etc)? Do they only originate from the overseer or from somewhere else as well?

Reply:
Exactly signals originate from the overseer and are prioritized.

The suggestion is to have an additional call recv_signal(), which only receives signals, but no messages ever. Indeed, we could make it three functions in total:

- recv ... the existing call, receiving everything.
- recv_signal ... only receiving signals.
- recv_msg .... only receiving messages.

Then you would either use recv or the recv_signal/recv_msg combination, giving you more control

Prioritizing signals even more than what is done by default should not do any harm. The overseer already ensures that you are never receiving messages that have been sent after a leaf activation, before you have seen that signal as well.

Reply:
> Exactly signals originate from the overseer and are prioritized.

Is it possible to document this @s0me0ne-unkn0wn? Also, the docs for `OverseerSignal` are not very helpful: "Signals sent by an overseer to a subsystem".

Reply:
Signals are sent as a form of broadcast to all subsystems by integration code, here it's (from memory) from the substrate block import pipeline (with a few intermediate steps).

Reply:
I think the better solution is to change orchestra to allow implementation of subsystems that don't really care about signals, maybe except `Conclude`. What I mean is that we either have a per subsystem configurable `SIGNAL_TIMEOUT`, or annotate these subsystems such that `broadcast_signal` only forwards them `Conclude`. Then we'd have to be careful to ensure nodes get stuck at shutdown because of this.